lib/gooddata/models/user_filters/user_filter_builder.rb

```diff
@@ -299,21 +302,42 @@ def self.create_user_profile_mapping(filters, project_users, options = {})
           missing_list << login
         end
       end
-
+      # rubocop:disable Metrics/BlockNesting
       unless missing_list.empty? || domain.nil?
         if missing_list.size < 100
           missing_list.each do |login|
             user = domain.find_user_by_login(login)
             found_list[login] = user.links['self'] if user
           end
         else
-          domain_users = domain.users
+          if @all_domain_users[domain.name].nil?
+            @mutex.lock
+            if @all_domain_users[domain.name].nil?
+              begin
+                domain_users = domain.users
+                @all_domain_users[domain.name] = domain_users
+                GoodData.logger.info("action=lcm_get_domain_users domain=#{domain.name} number_users=#{domain_users.size} missing_users=#{missing_list.size} use_cache=false")
+              rescue StandardError => e
+                GoodData.logger.warn("action=lcm_get_domain_users domain=#{domain.name} error=#{e}")
+              ensure
+                @mutex.unlock
```

Comment:
mutex is not unlocked if if @all_domain_users[domain.name] != nill?

Reply:
Only the first thread which is calling API and cached will do the lock/unlock. Other will wait for unlocking then used cached. So we don't need thread with @all_domain_users[domain.name] != nill? unlock the mutex. Thanks anh